### ISSUE TITLE
Minimum Instance Autoscaling

### ISSUE THREAD

**Issue by rsidhaarth** [vitobotta/hetzner-k3s#192]
In worker node pools, I have one set of autoscaling pods. I must have 4 nodes at the start and then auto scale further. But it is not working.

```
- name: autoscaling-node
  instance_type: ccx22
  instance_count: 4
  location: hel1
  autoscaling:
    enabled: true
    min_instances: 4
    max_instances: 25
```

But the auto scaler has deployed only one node.

**Comment by vitobotta** [comment]
Are you still having problems with this? Are you sure you haven't gone above the limit of the servers you can create with your account? You need to request a limit increase in that case.

**Comment by kim0** [comment]
pod anti-affinity rules might be helpful to you @rsidhaarth 
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity

**Comment by vitobotta** [comment]
> Yes, you can try to create a cluster with the above specs and try. It will deploy only 1, and then it will add one by one only when the pods need one.
> 
> This needs to be fixed because, if there is only one instance then all the pods get deployed in one instance only.

My question was about the Hetzner server limit. Each account has a limit to the number of servers it can creare before you need to request a limit increase. Are you sure you haven't reached this limit? This could explain why more nodes are not created. Also check the autoscaler logs to see if there is useful information there.

**Comment by vitobotta** [comment]
Did you check the autoscaler logs as I suggested?

**Comment by vitobotta** [comment]
I just tested with the latest version of the autoscaler and I can confirm that the min instances count seems to be ignored. The autoscaler only creates nodes when pods are pending. Not much I can do for now but I will leave this open as a reminder to check again with the next version. I will try to open an issue about it.

**Comment by clouedoc** [comment]
For information, the same issue arised to me with hetzner-k3s v1.1.2, when I tried configuring autoscaling ~1-2 days ago.

**Comment by sebastianklein96** [comment]
Indeed the autoscaler is showing strange behavior.

Deployed a cluster with the cluster config below using hetzner-k3s 1.1.3. I would have expected that three master nodes, one static worker node and 0 dynamic worker nodes would be deployed initially. But as soon as the cluster autoscaler is deployed by hetzner-k3s, a dynamic node in the dynamic node pool is provisioned. 

**Cluster Autoscaler log:**
```
I0630 08:53:14.818685       1 leaderelection.go:248] attempting to acquire leader lease kube-system/cluster-autoscaler...
I0630 08:53:14.919295       1 leaderelection.go:258] successfully acquired lease kube-system/cluster-autoscaler
W0630 08:53:16.932098       1 hetzner_servers_cache.go:94] Fetching servers from Hetzner API
I0630 08:53:17.118858       1 node_instances_cache.go:156] Start refreshing cloud provider node instances cache
I0630 08:53:17.119290       1 node_instances_cache.go:168] Refresh cloud provider node instances cache finished, refresh took 52.121µs
I0630 08:53:27.119379       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:53:27.119459       1 hetzner_node_group.go:438] Set node group dynamic-small size from 0 to 0, expected delta 0
W0630 08:53:27.119952       1 hetzner_server_type_cache.go:93] Fetching server types from Hetzner API
W0630 08:53:27.194532       1 clusterstate.go:428] AcceptableRanges have not been populated yet. Skip checking
W0630 08:53:27.194575       1 clusterstate.go:428] AcceptableRanges have not been populated yet. Skip checking
I0630 08:53:27.195973       1 scale_up.go:608] Scale-up: setting group dynamic-small size to 1
W0630 08:53:34.749221       1 hetzner_servers_cache.go:94] Fetching servers from Hetzner API
I0630 08:53:45.021089       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:53:45.021175       1 hetzner_node_group.go:438] Set node group dynamic-small size from 1 to 1, expected delta 0
I0630 08:53:55.045205       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:53:55.045311       1 hetzner_node_group.go:438] Set node group dynamic-small size from 1 to 1, expected delta 0
I0630 08:54:05.074732       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:54:05.074813       1 hetzner_node_group.go:438] Set node group dynamic-small size from 1 to 1, expected delta 0
I0630 08:54:15.092047       1 hetzner_node_group.go:438] Set node group dynamic-small size from 1 to 1, expected delta 0
I0630 08:54:15.092419       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:54:25.124240       1 hetzner_node_group.go:438] Set node group draining-node-pool size from 0 to 0, expected delta 0
I0630 08:54:25.124387       1 hetzner_node_group.go:438] Set node group dynamic-small size from 1 to 1, expected delta 0
```


**Cluster Config:**
```
hetzner_token: *redacted*
cluster_name: *redacted*
kubeconfig_path: "./kubeconfig"
k3s_version: v1.26.4+k3s1
public_ssh_key_path: "~/.ssh/id_rsa.pub"
private_ssh_key_path: "~/.ssh/id_rsa"
use_ssh_agent: false
ssh_allowed_networks:
  - 0.0.0.0/0
api_allowed_networks:
  - 0.0.0.0/0
schedule_workloads_on_masters: false
private_network_subnet: 10.9.0.0/16
disable_flannel: false # set to true if you want to install a different CNI
cloud_controller_manager_manifest_url: "https://github.com/hetznercloud/hcloud-cloud-controller-manager/releases/download/v1.16.0/ccm-networks.yaml"
csi_driver_manifest_url: "https://raw.githubusercontent.com/hetznercloud/csi-driver/v2.3.2/deploy/kubernetes/hcloud-csi.yml"
system_upgrade_controller_manifest_url: "https://raw.githubusercontent.com/rancher/system-upgrade-controller/master/manifests/system-upgrade-controller.yaml"
masters_pool:
  instance_type: cx21
  instance_count: 3
  location: fsn1
worker_node_pools:
- name: static
  instance_type: cx31
  instance_count: 1
  location: fsn1
- name: dynamic-small
  instance_type: cx31
  instance_count: 0
  location: fsn1
  autoscaling:
    enabled: true
    min_instances: 0
    max_instances: 5
additional_packages:
- wireguard
post_create_commands:
- apt update
- apt upgrade -y
- apt autoremove -y
```

**Comment by vitobotta** [comment]
@sebastianklein96 What did you deploy in the cluster and with which resource requests? If the autoscaler has created a node it's likely because there were one or more pods that could not be scheduled for lack of resources.

**Comment by sebastianklein96** [comment]
That is what's confusing me. This happens with a completely fresh cluster, during deployment using hetzner-k3s. I am not deploying any workloads at all. The dynamic worker is created immediately after deployment of the cluster autoscaler. So in theory there should not be a lack of resources. This also happens when deploying more than one static worker.

**Comment by vitobotta** [comment]
Weird, it has never happened to me. Perhaps something has changed with the latest version of the autoscaler. I will give it a try as soon as I have some time.

**Comment by manfredipist** [comment]
I can confirm that the behaviour of the autoscaler has changed lately. I've got 2 clusters, one deployed 3 months ago on v1.25.3 and another one deployed just yesterday on v 1.26.4. In my previous cluster the autoscaler is only getting triggered when pods cannot be scheduled, without respecting my will about desired minInstances. On my latest cluster, on the other hand, the autoscaler is deploying my desired amount of "minIstances"

**Comment by vitobotta** [comment]
> I can confirm that the behaviour of the autoscaler has changed lately. I've got 2 clusters, one deployed 3 months ago on v1.25.3 and another one deployed just yesterday on v 1.26.4. In my previous cluster the autoscaler is only getting triggered when pods cannot be scheduled, without respecting my will about desired minInstances. On my latest cluster, on the other hand, the autoscaler is deploying my desired amount of "minIstances"

Thanks, that sounds promising. I will do some testing once I have some time.

**Comment by vitobotta** [comment]
I released v1.1.4 and tested with autoscaler 1.27.something and still ignores the minimum instance count. :(

**Comment by vitobotta** [comment]
I just found out that this behaviour is expected so I am closing this issue since I can't do anything about it.

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#my-cluster-is-below-minimum--above-maximum-number-of-nodes-but-ca-did-not-fix-that-why

>  If your cluster is below the minimum number of nodes configured for CA, it will be scaled up only in presence of unschedulable pods.
